src/global.c

```diff
@@ -27,153 +27,32 @@
-
-static git_atomic git__n_shutdown_callbacks;
-static git_atomic git__n_inits;
-
-void git__on_shutdown(git_global_shutdown_fn callback)
-{
-	int count = git_atomic_inc(&git__n_shutdown_callbacks);
-	assert(count <= (int) ARRAY_SIZE(git__shutdown_callbacks) && count > 0);
-	git__shutdown_callbacks[count - 1] = callback;
-}
-
-static int init_common(void)
-{
-	size_t i;
-	int ret;
-
-	/* Initialize subsystems that have global state */
-	for (i = 0; i < ARRAY_SIZE(git__init_callbacks); i++)
-		if ((ret = git__init_callbacks[i]()) != 0)
-			break;
-
-	GIT_MEMORY_BARRIER;
-
-	return ret;
-}
-
-static void shutdown_common(void)
-{
-	int pos;
-
-	/* Shutdown subsystems that have registered */
-	for (pos = git_atomic_get(&git__n_shutdown_callbacks);
-		pos > 0;
-		pos = git_atomic_dec(&git__n_shutdown_callbacks)) {
-
-		git_global_shutdown_fn cb = git__swap(
-			git__shutdown_callbacks[pos - 1], NULL);
-
-		if (cb != NULL)
-			cb();
-	}
-}
-
-/*
- * `git_libgit2_init()` allows subsystems to perform global setup,
- * which may take place in the global scope.  An explicit memory
- * fence exists at the exit of `git_libgit2_init()`.  Without this,
- * CPU cores are free to reorder cache invalidation of `_tls_init`
- * before cache invalidation of the subsystems' newly written global
- * state.
- */
-#if defined(GIT_THREADS) && defined(GIT_WIN32)
-
-static volatile LONG init_mutex = 0;
-
-GIT_INLINE(int) mutex_lock(void)
-{
-	while (InterlockedCompareExchange(&init_mutex, 1, 0)) { Sleep(0); }
-	return 0;
-}
-
-GIT_INLINE(int) mutex_unlock(void)
-{
-	InterlockedExchange(&init_mutex, 0);
-	return 0;
-}
-
-#elif defined(GIT_THREADS) && defined(_POSIX_THREADS)
-
-static pthread_mutex_t init_mutex = PTHREAD_MUTEX_INITIALIZER;
-
-GIT_INLINE(int) mutex_lock(void)
-{
-	return pthread_mutex_lock(&init_mutex) == 0 ? 0 : -1;
-}
-
-GIT_INLINE(int) mutex_unlock(void)
-{
-	return pthread_mutex_unlock(&init_mutex) == 0 ? 0 : -1;
-}
-
-#elif defined(GIT_THREADS)
-# error unknown threading model
-#else
-
-# define mutex_lock() 0
-# define mutex_unlock() 0
-
-#endif
-
 int git_libgit2_init(void)
 {
-	int ret;
-
-	if (mutex_lock() < 0)
-		return -1;
-
-	/* Only do work on a 0 -> 1 transition of the refcount */
-	if ((ret = git_atomic_inc(&git__n_inits)) == 1) {
-		if (init_common() < 0)
-			ret = -1;
-	}
-
-	if (mutex_unlock() < 0)
-		return -1;
-
-	return ret;
+	static git_runtime_init_fn init_fns[] = {
+#if defined(GIT_MSVC_CRTDBG)
+		git_win32__crtdbg_stacktrace_init,
+		git_win32__stack_init,
+#endif
+		git_allocator_global_init,
+		git_tlsdata_global_init,
+		git_threads_global_init,
+		git_hash_global_init,
+		git_sysdir_global_init,
+		git_filter_global_init,
+		git_merge_driver_global_init,
+		git_transport_ssh_global_init,
+		git_stream_registry_global_init,
+		git_openssl_stream_global_init,
+		git_mbedtls_stream_global_init,
+		git_mwindow_global_init,
+		git_settings_global_init
+	};
+
+	return git_runtime_init(init_fns, ARRAY_SIZE(init_fns));
```

Comment:
Hum. You're probably keeping "global.c" and "runtime.c" separate in order to not leak all the different headers into "global.c"? I'd have just moved all of "global.c" into "runtime.c", mostly because I'm not much of a fan of the name "global" and think that "runtime" is a much better name. Given that we now only have `git_libgit2_init()` and `git_libgit2_shutdown()` left, we could also rename this file to "libgit2.c" instead.

Reply:
I moved this to libgit2.c, and moved all the `git_libgit2_` functions into that file for our understanding.